h5py/h5t.pyx

```diff
@@ -1712,6 +1712,42 @@ cpdef TypeID py_create(object dtype_in, bint logical=0, bint aligned=0):
         else:
             raise TypeError("No conversion path for dtype: %s" % repr(dt))
 
+def vlen_dtype(basetype):
+    """Make a numpy dtype for an HDF5 variable-length datatype
+
+    For variable-length string dtypes, use :func:`string_dtype` instead.
+    """
+    return dtype('O', metadata={'vlen': basetype})
+
+def string_dtype(encoding='utf-8'):
+    """Make a numpy dtype for HDF5 variable-length strings
+
+    encoding may be 'utf-8' or 'ascii'. If it is 'utf-8', the data should be
+    passed as Python str objects (unicode in Python 2). For 'ascii', data should
+    be passed as bytes.
+    """
+    if encoding == 'utf-8':
+        return dtype('O', metadata={'vlen': unicode})
+    elif encoding == 'ascii':
+        return dtype('O', metadata={'vlen': bytes})
+    else:
+        raise ValueError("Invalid encoding (%r); 'utf-8' or 'ascii' allowed"
+                         % encoding)
+
+def enum_dtype(values_dict, basetype=np.uint8):
+    """Create a NumPy representation of an HDF5 enumerated type
+
+    *values_dict* maps string names to integer values. *basetype* is an
+    appropriate integer base dtype large enough to hold the possible options.
+    """
+    dt = dtype(basetype)
+    if not np.issubdtype(dt, np.integer):
+        raise TypeError("Only integer types can be used as enums")
+
+    return dtype(dt, metadata={'enum': values_dict})
+
+ref_dtype = dtype('O', metadata={'ref': Reference})
```

Comment:
There is now one instance each of these two dtypes that are re-used, where as previously we would call `dytpe` each time.  Do we care and are `dtype` instance mutable?

Reply:
I'm pretty sure `dtype` instances are indeed mutable, for better or worse.

Reply:
That seems like a bit of a foot-cannon to hand to users....

Reply:
Do you want these to be functions returning new dtype objects on each call?

Reply:
I am not sure I have a good enough grasp of the trade-offs here. 

A function makes it more like the others and gets out of mutablity issue.

On the other hand, if dtypes are mutable then this also applies to the dtypes from numpy it's self which no one seems to worry about so we should not either and it is nice that these are constants.

Reply:
Maybe I was wrong here -- it seems that you cannot assign into the metadata dict after creating a dtype:
```
>>> d1 = np.dtype(object, metadata={'x': 1})
>>> d1.metadata['x'] = 2
TypeError: 'mappingproxy' object does not support item assignment
```

Reply:
And attempting to modify `dt.kind` and `dt.itemsize` gives me `AttributeError: readonly attribute
`. So maybe even if they're [not totally immutable](https://github.com/numpy/numpy/issues/1725), they're hard enough to modify that we needn't worry about the potential footgun.